File: src-tauri/crates/schema_types/src/checksum.rs
//! 校验算法 — CRC / SUM / XOR / LRC 等。
//!
//! 与前端 `ChecksumType` 对齐, serde rename 显式指定字符串。

use serde::{Deserialize, Serialize};

/// 校验算法 (与前端 ChecksumType 对齐, serde rename 显式指定字符串)
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum ChecksumAlgorithm {
    #[serde(rename = "none")]
    None,
    #[serde(rename = "sum8")]
    Sum8,
    #[serde(rename = "xor8")]
    Xor8,
    #[serde(rename = "crc8")]
    Crc8,
    #[serde(rename = "crc16Modbus")]
    Crc16Modbus,
    #[serde(rename = "crc16CCITT")]
    Crc16CCITT,
    #[serde(rename = "crc32")]
    Crc32,
    #[serde(rename = "lrc")]
    Lrc,
    #[serde(rename = "custom")]
    Custom,
}

impl ChecksumAlgorithm {
    /// 计算校验值 (返回单字节或 4 字节, 由调用方截取)
    pub fn compute(self, data: &[u8], custom_script: Option<&str>) -> Vec<u8> {
        match self {
            Self::None => Vec::new(),
            Self::Sum8 => {
                let s: u8 = data.iter().fold(0u8, |acc, &b| acc.wrapping_add(b));
                vec![s]
            }
            Self::Xor8 => {
                let x: u8 = data.iter().fold(0u8, |acc, &b| acc ^ b);
                vec![x]
            }
            Self::Crc8 => vec![crc8(data, 0x07, 0x00, 0x00)],
            Self::Crc16Modbus => {
                let crc = crc16_modbus(data);
                crc.to_le_bytes().to_vec()
            }
            Self::Crc16CCITT => {
                let crc = crc16_ccitt(data);
                crc.to_be_bytes().to_vec()
            }
            Self::Crc32 => {
                let crc = crc32(data);
                crc.to_le_bytes().to_vec()
            }
            Self::Lrc => {
                let lrc: u8 = data.iter().fold(0u8, |acc, &b| acc.wrapping_sub(b));
                vec![lrc]
            }
            Self::Custom => {
                // 自定义脚本暂不支持后端求值 (前端 lib/checksum.ts 中的 customChecksum 用 JS 实现)
                // 后端此处返回空 Vec, 实际项目应引入 rhai/boa 等 JS 引擎求值
                let _ = custom_script;
                Vec::new()
            }
        }
    }

    /// 比较计算值与期望值 (自动处理长度差异)
    pub fn verify(self, data: &[u8], expected: &[u8], custom_script: Option<&str>) -> bool {
        let computed = self.compute(data, custom_script);
        if computed.is_empty() {
            return true; // None / Custom 未实现 → 默认通过
        }
        computed == expected
    }

    /// 校验算法输出的字节长度
    pub const fn byte_len(self) -> usize {
        match self {
            Self::None => 0,
            Self::Sum8 | Self::Xor8 | Self::Crc8 | Self::Lrc => 1,
            Self::Crc16Modbus | Self::Crc16CCITT => 2,
            Self::Crc32 => 4,
            Self::Custom => 0, // Custom 暂不支持后端求值
        }
    }
}

/// CRC-8 (poly=0x07, init=0x00, refin=false, refout=false, xorout=0x00)
fn crc8(data: &[u8], poly: u8, init: u8, xorout: u8) -> u8 {
    let mut crc = init;
    for &b in data {
        crc ^= b;
        for _ in 0..8 {
            if crc & 0x80 != 0 {
                crc = (crc << 1) ^ poly;
            } else {
                crc <<= 1;
            }
        }
    }
    crc ^ xorout
}

/// CRC-16 Modbus (poly=0x8005, init=0xFFFF, refin=true, refout=true, xorout=0x0000)
fn crc16_modbus(data: &[u8]) -> u16 {
    let mut crc: u16 = 0xFFFF;
    for &b in data {
        crc ^= u16::from(b);
        for _ in 0..8 {
            if crc & 1 != 0 {
                crc = (crc >> 1) ^ 0xA001; // 0x8005 反转
            } else {
                crc >>= 1;
            }
        }
    }
    crc
}

/// CRC-16 CCITT (poly=0x1021, init=0xFFFF, refin=false, refout=false, xorout=0x0000)
fn crc16_ccitt(data: &[u8]) -> u16 {
    let mut crc: u16 = 0xFFFF;
    for &b in data {
        crc ^= u16::from(b) << 8;
        for _ in 0..8 {
            if crc & 0x8000 != 0 {
                crc = (crc << 1) ^ 0x1021;
            } else {
                crc <<= 1;
            }
        }
    }
    crc
}

/// CRC-32 (poly=0x04C11DB7, init=0xFFFFFFFF, refin=true, refout=true, xorout=0xFFFFFFFF)
fn crc32(data: &[u8]) -> u32 {
    let mut crc: u32 = 0xFFFFFFFF;
    for &b in data {
        crc ^= u32::from(b);
        for _ in 0..8 {
            if crc & 1 != 0 {
                crc = (crc >> 1) ^ 0xEDB88320; // 0x04C11DB7 反转
            } else {
                crc >>= 1;
            }
        }
    }
    !crc
}

File: src/lib/utils/checksum.ts
/// 校验和计算库 — 各类嵌入式通信常用校验算法
///
/// 所有函数输入为 Uint8Array, 输出为 number[] (1~4 字节, 按高位在前/低位在前视算法而定)
///
/// 算法参考:
/// - CRC: https://reveng.sourceforge.io/crc-catalogue/
/// - LRC: Modbus Over Serial Line — LRC 计算

/// CRC-8 (poly 0x07, init 0x00, reflIn/reflOut=false, xorOut=0x00)
/// 用于 SMBus / 一般 8-bit 校验
export function crc8(data: Uint8Array): number[] {
  let crc = 0x00;
  for (const b of data) {
    crc ^= b;
    for (let i = 0; i < 8; i++) {
      crc = (crc & 0x80) ? ((crc << 1) ^ 0x07) & 0xff : (crc << 1) & 0xff;
    }
  }
  return [crc];
}

/// CRC-16 Modbus (poly 0xA001, init 0xFFFF, reflIn/reflOut=true, xorOut=0x0000)
/// 用于 Modbus RTU
export function crc16Modbus(data: Uint8Array): number[] {
  let crc = 0xffff;
  for (const b of data) {
    crc ^= b;
    for (let i = 0; i < 8; i++) {
      crc = (crc & 0x0001) ? ((crc >> 1) ^ 0xa001) & 0xffff : (crc >> 1) & 0xffff;
    }
  }
  // Modbus 低字节在前
  return [crc & 0xff, (crc >> 8) & 0xff];
}

/// CRC-16 CCITT-FALSE (poly 0x1021, init 0xFFFF, reflIn/reflOut=false, xorOut=0x0000)
/// 用于 XMODEM / 一般通信帧
export function crc16CCITT(data: Uint8Array): number[] {
  let crc = 0xffff;
  for (const b of data) {
    crc ^= (b << 8);
    for (let i = 0; i < 8; i++) {
      crc = (crc & 0x8000) ? ((crc << 1) ^ 0x1021) & 0xffff : (crc << 1) & 0xffff;
    }
  }
  // 高字节在前 (big-endian)
  return [(crc >> 8) & 0xff, crc & 0xff];
}

/// CRC-32 (ZIP poly 0xEDB88320 反射, init 0xFFFFFFFF, xorOut=0xFFFFFFFF)
/// 用于 ZIP / Ethernet
export function crc32(data: Uint8Array): number[] {
  let crc = 0xffffffff;
  for (const b of data) {
    crc ^= b;
    for (let i = 0; i < 8; i++) {
      crc = (crc & 1) ? ((crc >>> 1) ^ 0xedb88320) >>> 0 : (crc >>> 1);
    }
  }
  crc ^= 0xffffffff;
  // 低字节在前 (little-endian)
  return [
    crc & 0xff,
    (crc >>> 8) & 0xff,
    (crc >>> 16) & 0xff,
    (crc >>> 24) & 0xff,
  ];
}

/// Sum8 — 累加和 & 0xFF (校验和)
export function sum8(data: Uint8Array): number[] {
  let s = 0;
  for (const b of data) s = (s + b) & 0xff;
  return [s];
}

/// XOR8 — 逐字节异或
export function xor8(data: Uint8Array): number[] {
  let x = 0;
  for (const b of data) x ^= b;
  return [x];
}

/// LRC — Modbus ASCII Longitudinal Redundancy Check
/// 两倍补码和的低字节
export function lrc(data: Uint8Array): number[] {
  let s = 0;
  for (const b of data) s = (s + b) & 0xff;
  return [((-s) & 0xff)];
}

/// 自定义校验脚本 — 用户提供的 JS 函数体, 接收 bytes 数组, 返回 number[]
///
/// 警告: 使用 new Function 动态求值, 有安全风险 (运行在主进程), 性能不如内置算法
///
/// 示例脚本:
///   // 简单 sum + xor
///   let s = 0, x = 0;
///   for (const b of bytes) { s = (s + b) & 0xff; x ^= b; }
///   return [s, x];
///
/// 沙箱限制: 仅能访问 bytes 参数, 不能访问 window/document/eval
export function customChecksum(data: Uint8Array, script: string): number[] {
  try {
    // Custom checksum snippets are an explicit user-configurable extension point.
    // eslint-disable-next-line @typescript-eslint/no-implied-eval
    const fn = new Function('bytes', script) as (bytes: number[]) => number[];
    const result = fn(Array.from(data));
    if (!Array.isArray(result)) {
      throw new Error('Custom checksum script must return an array of numbers');
    }
    return result.map((n) => {
      const num = Number(n);
      if (!Number.isFinite(num)) throw new Error(`Invalid number: ${n}`);
      return num & 0xff;
    });
  } catch (e) {
    const wrapped = new Error(`Custom checksum script error: ${(e as Error).message}`);
    Object.defineProperty(wrapped, 'cause', { value: e, configurable: true });
    throw wrapped;
  }
}

/// 校验类型枚举 (与 types/index.ts 的 ChecksumType 对应)
export type ChecksumKind =
  | 'none'
  | 'sum8'
  | 'xor8'
  | 'crc8'
  | 'crc16Modbus'
  | 'crc16CCITT'
  | 'crc32'
  | 'lrc'
  | 'custom';

/// 计算指定类型的校验和
export function computeChecksum(
  data: Uint8Array,
  kind: ChecksumKind,
  customScript?: string
): number[] {
  switch (kind) {
    case 'none': return [];
    case 'sum8': return sum8(data);
    case 'xor8': return xor8(data);
    case 'crc8': return crc8(data);
    case 'crc16Modbus': return crc16Modbus(data);
    case 'crc16CCITT': return crc16CCITT(data);
    case 'crc32': return crc32(data);
    case 'lrc': return lrc(data);
    case 'custom':
      if (!customScript) return [];
      return customChecksum(data, customScript);
  }
}

/// 校验和字节长度 (用于显示)
export function checksumLength(kind: ChecksumKind): number {
  switch (kind) {
    case 'none': return 0;
    case 'crc32': return 4;
    case 'crc16Modbus':
    case 'crc16CCITT': return 2;
    case 'sum8':
    case 'xor8':
    case 'crc8':
    case 'lrc':
    case 'custom': return 1; // 自定义可能多字节, 显示为 1 (实际看脚本)
  }
}
